internal/cli/atlas/backup/compliancepolicy/watcher/watcher.go

```diff
@@ -0,0 +1,49 @@
+// Copyright 2023 MongoDB Inc
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package watcher
+
+import (
+	"errors"
+
+	store "github.com/mongodb/mongodb-atlas-cli/internal/store/atlas"
+	atlasv2 "go.mongodb.org/atlas-sdk/v20230201004/admin"
+)
+
+const (
+	active = "ACTIVE"
+)
+
+var errInvalidStateField = errors.New("invalid State field for watching")
+
+// CompliancePolicyWatcherFactory allows to init a compliance policy watcher in a functional way.
+//
+// Parameters
+//
+//   - projectID: The ID of the project for which the compliance policy details need to be fetched.
+//   - store: An implementation of the CompliancePolicyDescriber interface, which is used to describe/fetch the compliance policy.
+//   - policy: A pointer to a DataProtectionSettings object which will be updated with the fetched details.
+func CompliancePolicyWatcherFactory(projectID string, store store.CompliancePolicyDescriber, policy *atlasv2.DataProtectionSettings) func() (bool, error) {
```

Comment:
any reason to not go for a more common pattern in the cli of composition?

have you considered
```go
type polictyOpts struct {
	policy          *atlasv2.DataProtectionSettings
	store           store.CompliancePolicy
}

func (opts *polictyOpts) watcher(projectID string) (bool, error) {
	res, err := opts.store.DescribeCompliancePolicy(projectID)
	if err != nil {
		return false, err
	}
	opts.policy = res
	if res.GetState() == "" {
		return false, fmt.Errorf("could not access State field")
	}
	return res.GetState() == active, nil
}
```

Reply:
The reason I did not go for this pattern was to avoid introducing more state. I don't have a strong opinion either way, the question for me is commonality vs state. If you have a stronger opinion on this, I'll change it accordingly.

Reply:
the state already exists, you'll be lifting it to a common struct

Reply:
the stronger opinion I have right now is your refactor doesn't really save much except for go bolierplate code and according to the [go proverbs](https://go-proverbs.github.io/)

[A little copying is better than a little dependency.](https://www.youtube.com/watch?v=PAAkCSZUG1c&t=9m28s)

Reply:
Also notice this function currently modifies an input parameter which in my opinion is worst than state

Reply:
for example 
```go
// CompliancePolicyState encapsulates the logic to manage different cloud providers.
func (s *Store) CompliancePolicyState(projectID string) (*atlasv2.DataProtectionSettings, string, error) {
	switch s.service {
	case config.CloudService, config.CloudGovService:
		result, _, err := s.clientv2.CloudBackupsApi.GetDataProtectionSettings(s.ctx, projectID).Execute()
		if err != nil {
			return result, "", err
		}
		if result.GetState() == "" {
			return result, "", errInvalidStateField
		}		
		return result, result.GetState(), err
	default:
		return nil, "", fmt.Errorf("%w: %s", errUnsupportedService, s.service)
	}
}
```

Reply:
> the stronger opinion I have right now is your refactor doesn't really save much except for go bolierplate code and according to the [go proverbs](https://go-proverbs.github.io/)
> 
> [A little copying is better than a little dependency.](https://www.youtube.com/watch?v=PAAkCSZUG1c&t=9m28s)

I agree. Is it ok with you if I close the PR?

Reply:
up to you, these are the options in no preferred order
- close the PR
- uplift the state to a common struct all sub-commands can include
- implement the checks on the store